### PR TITLE
Updated to support API changes for Mambu 3.14

### PR DESCRIPTION
**Pull request by mdanilkis**
Major additions:
- Update Savings Account Terms API. See MBU-10447
- GET User Roles and GET User Role APIs. See MBU-9263 and MBU-9263
- Write Off Loan Accounts API. See MBU-10423
- GET Document Templates API: GET Populated Account and Transaction Templates. See MBU-10631) and  MBU-10632 . Added DocumentTemplatesService
- GET  loan accounts funded by a deposit investor account. See MBU-10905
- GET Entities by Custom View API now supports BASIC and FULL_DETAILS  result types. See MBU-10842. Added CustomViewsService.
- Added missing methods for GET GL accounts and GET Journal Entries. Available since v2.0. See MBU-1543 and MBU-1736
- Added missing (since 3.6) methods to GET Documents for Products, Users, Branches and Centres. Enhanced existent methods to support pagination. See MBU-5084

Changes:
- Communication Messages History can be now searched via On The Fly filtering APIs.  See MBU-10646
- Repayments Schedule API:  added support for Fixed Days instalments (which now can be specified at the account level). See MBU-10205

Miscellaneous Enhancements:
DemoTests:
a) added support for running regression tests for accounts in a loop for all supported product types
b) added support for testing new Revolving Credit product types. See MBU-10111, MBU-10546
c) testing support for On the Fly Search API:  Sorting Columns (MBU-10444) and filter entities based on Another Entity criteria (MBU-8985).
d) added support for Search API to test searching Lines of Credit.  New in 3.14

Updated to support 3.14 model changes. Example, MBU-9725,  MBU-10949

### CODE REVIEW

File: src/com/mambu/apisdk/model/ApiGLJournalEntry.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+	public void setGlCode(String glCode) {
+		this.glCode = glCode;
+	}
+
+	public void setEntryType(EntryType entryType) {
+		this.entryType = entryType;
+	}
+
+	public void setAmount(String amount) {
```
Comment (thobach) [inline src/com/mambu/apisdk/model/ApiGLJournalEntry.java]: I know the API expects a JSON string here, but since you added a wrapper class, does it make sense to accept BigDecimal here?

File: src/com/mambu/apisdk/services/AccountingService.java [hunk 2]
```diff
@@ -89,17 +127,22 @@ public GLAccount getGLAccount(String glCode, String fromDate, String toDate) thr
 	 * @throws MambuApiException
 	 *             in case of an error
 	 */
-	public List<GLJournalEntry> getGLJournalEntries(Date fromDate, Date toDate) throws MambuApiException {
```
Comment (thobach) [inline src/com/mambu/apisdk/services/AccountingService.java]: Please don't remove public methods without deprecating them first.

File: src/com/mambu/apisdk/services/AccountingService.java [hunk 3]
```diff
@@ -110,14 +153,16 @@ public GLAccount getGLAccount(String glCode, String fromDate, String toDate) thr
 	 * @throws MambuApiException
 	 *             in case of an error
 	 */
-	public List<GLJournalEntry> getGLJournalEntries(Date fromDate, Date toDate, int offset, int limit)
```
Comment (thobach) [inline src/com/mambu/apisdk/services/AccountingService.java]: Please don't remove public methods without deprecating them first.

File: src/com/mambu/apisdk/services/AccountingService.java [hunk 4]
```diff
@@ -125,4 +170,79 @@ public GLAccount getGLAccount(String glCode, String fromDate, String toDate) thr
 
 		return serviceExecutor.execute(getGLJournalEntries, params);
 	}
+
+	// Params definitions for posting GL Journal Entries. See MBU-1737
+	private static String debitAccountParam = "debitAccount";
```
Comment (thobach) [inline src/com/mambu/apisdk/services/AccountingService.java]: Please move these static fields up to the top.

File: src/com/mambu/apisdk/services/AccountingService.java [hunk 5]
```diff
@@ -125,4 +170,79 @@ public GLAccount getGLAccount(String glCode, String fromDate, String toDate) thr
...
+	 *            The date of the posting of the journal entry. Must be not null
+	 * @param notes
+	 *            transaction notes
+	 * @return created journal entries
+	 * @throws MambuApiException
+	 */
+	public List<GLJournalEntry> postGLJournalEntries(List<ApiGLJournalEntry> entries, String branchId, String date,
+			String notes) throws MambuApiException {
+		// POST
+		// api/gljournalentries/branchId=b01&date=2012-02-05&debitAccount1=1003&debitAmount1=-0.05&creditAccount1=1003&creditAmount1=-0.05"
```
Comment (thobach) [inline src/com/mambu/apisdk/services/AccountingService.java]: Are the parameters added to the URL or posted in the body? They should be part of the body.

File: src/com/mambu/apisdk/services/AccountingService.java [hunk 6]
```diff
@@ -125,4 +170,79 @@ public GLAccount getGLAccount(String glCode, String fromDate, String toDate) thr
...
+		// See MBU-1737
+		if (entries == null || entries.size() == 0) {
+			throw new IllegalArgumentException("At least one entry is required");
+		}
+		if (date == null) {
+			throw new IllegalArgumentException("Date must not be null");
+		}
+
+		ParamsMap params = new ParamsMap();
+		int index = 0;
```
Comment (thobach) [inline src/com/mambu/apisdk/services/AccountingService.java]: In the examples on the Jira issue the index always starts with 1, does it also work with 0?
I don't see the index being increased, could you double check that and add a test case also in case none exists yet?

File: src/com/mambu/apisdk/model/ApiGLJournalEntry.java (hunk 1 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/model/ApiGLJournalEntry.java]: Yes, it does make good sense. I will change the amount to be a BigDecimal instead of a String

File: src/com/mambu/apisdk/services/AccountingService.java (hunk 2 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/AccountingService.java]: Right.  I will add back methods with no "branchId" parameter and will deprecate them

File: src/com/mambu/apisdk/services/AccountingService.java (hunk 4 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/AccountingService.java]: OK. I will move them to the APIData class (where all other simmer parameters are)

File: src/com/mambu/apisdk/services/AccountingService.java (hunk 5 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/AccountingService.java]: Yes, all parameters are part of the body (this method uses Method.POST, so they will be posted in the body as for all POSTs). (The documentation example in this method was copied from the original Mambu issue description, I will update it as well for clarity)

File: src/com/mambu/apisdk/services/AccountingService.java (hunk 6 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/AccountingService.java]: Good catch! The index should be increased. I will fix this. There are already demo tests for this method but there was only one debit/credit pair in a demo test data, so this bug wasn't picked up.  I will update the demo tests as well to test POSTing multiple transactions.
The index works both when starting from zero or starting from one. To match API documentation examples I will change it to  start from 1.

**Comment by thobach** [comment]
Thanks a lot for all these additions and improvements! I finished the quick review (skipped reviewing demo code changes).
